hack/release.sh

```diff
@@ -0,0 +1,200 @@
+
+if [ -z "$ALLOW_DIRTY" -a "$GIT_DIRTY" != "clean" ]; then
+    echo "error: repo status is not clean, skipped"
+    exit 1
+fi
+
+# our logic depends repo tags, make sure all tags are fetched
+echo "info: fetching all tags"
+git fetch --tags
+
+if [ -z "$VERSION" ]; then
+    echo "info: VERSION is not specified, detect automatically"
+    # get version from tag
+    VERSION=$(git describe --tags --abbrev=0 --exact-match 2>/dev/null || true)
+    if [ -z "$VERSION" ]; then
+        echo "error: failed to detect version"
+        exit 1
+    fi
+fi
+
+# By default, only v<major>.<minor>.<patch> version are allowed.
+function is_stable_version() {
+    [[ "$1" =~ ^v[0-9]+\.[0-9]+\.[0-9]+$ ]]
+}
+
+if [ -z "$ALLOW_UNSTABLE" ]; then
+    if ! is_stable_version "$VERSION"; then
+        echo "error: unstable version '$VERSION', must match regex 'v[0-9]+\.[0-9]+\.[0-9]+$' or skip this check with ALLOW_UNSTABLE"
+        exit 1
+    fi
+fi
+
+image="$REGISTRY/local-volume-provisioner:$VERSION"
+if [ -z "$CONFIRM" ]; then
+    read -r -p "info: build and push to $image? [y/N]" response
+    if [[ ! $response =~ ^([yY][eE][sS]|[yY])$ ]]; then
+        echo "Exited."
+        exit 0
+    fi
+fi
+
+# Usage: manifest_url <domain>/<name>:<reference>
+# xref: https://docs.docker.com/registry/spec/api/
+function get_manifest_url() {
+    read -r domain namewithref <<<"${1/\// }"
+    read -r name ref <<<"${namewithref/:/ }"
+    echo "https://${domain}/v2/${name}/manifests/${ref}"
+}
+
+if [ -z "$ALLOW_OVERRIDE" ] && is_stable_version "$VERSION"; then
+    echo "info: $image is stable image, checking if it does eixst"
+    manifest_url=$(get_manifest_url "$image")
+    code=$(curl -s -XHEAD -w '%{http_code}' "$manifest_url" || true)
+    if [ "$code" == "200" ]; then
+        echo "error: $image does exist, skipped"
+        exit 1
+    elif [ "$code" != "404" ]; then
+        echo "error: unexpected http code '$code'"
+        exit 1
+    fi
+    echo "info: '$image' does not exist, continue"
+fi
+
+# build & push container
+
+if [ -z "$SKIP_BUILD" ]; then
+    echo "info: building $image"
+    pushd provisioner &>/dev/null
+    make container REGISTRY=$REGISTRY VERSION=$VERSION
+    popd
+else
+    echo "info: building is skipped"
+fi
+
+echo "info: pushing $image"
+docker_args=()
+if [ -n "$DOCKER_CONFIG" ]; then
+    if [ ! -d "$DOCKER_CONFIG" ]; then
+        echo "error: DOCKER_CONFIG '$DOCKER_CONFIG' does not exist or not a directory"
+        exit 1
+    fi
+    if [ ! -f "$DOCKER_CONFIG/config.json" ]; then
+        echo "error: docker config json '$DOCKER_CONFIG/config.json' does not exist"
+        exit 1
+    fi
+    docker_args+=(--config "$DOCKER_CONFIG")
+fi
+docker_args+=(push "$image")
+docker "${docker_args[@]}"
+
+function is_latest_version() {
+    local v="$1"
+}
+
+if ! is_stable_version "$VERSION"; then
+    echo "error: VERSION '$VERSION' is not stable version, skipped pushing as latest image"
+    exit 1
```

Comment:
do we want to exit with error here?  will this cause canary build jobs to fail?

Reply:
You are right. Job will fail, I will update it when configuring job in prow.